ipa-core/src/protocol/context/dzkp_field.rs

```diff
@@ -12,17 +25,332 @@ pub trait DZKPCompatibleField: Field {
-        x_left: &BitArray32,
-        x_right: &BitArray32,
-        y_left: &BitArray32,
-        y_right: &BitArray32,
-        prss_left: &BitArray32,
-        prss_right: &BitArray32,
-        z_right: &BitArray32,
-    ) -> Self::UnverifiedFieldValues;
+pub trait DZKPBaseField: PrimeField {
+    const INVERSE_OF_TWO: Self;
+    const MINUS_ONE_HALF: Self;
+    const MINUS_TWO: Self;
+
+    /// Convert allows to convert individual bits from multiplication gates into dzkp compatible field elements.
+    /// This function is called by the prover.
+    fn convert_prover<'a>(
+        x_left: &'a BitArray32,
+        x_right: &'a BitArray32,
+        y_left: &'a BitArray32,
+        y_right: &'a BitArray32,
+        prss_right: &'a BitArray32,
+    ) -> impl Iterator<Item = UVPolynomials<Self>>;
+
+    /// This is similar to `convert_prover` except that it is called by the verifier to the left.
+    /// The verifier on the left uses its right shares, since they are consistent with the prover's left shares.
+    fn convert_verifier_left<'a>(
+        x_right: &'a BitArray32,
+        y_right: &'a BitArray32,
+        prss_right: &'a BitArray32,
+        z_right: &'a BitArray32,
+    ) -> impl Iterator<Item = SingleUVPolynomial<Self>>;
+
+    /// This is similar to `convert_prover` except that it is called by the verifier to the right.
+    /// The verifier on the right uses its left shares, since they are consistent with the prover's right shares.
+    fn convert_verifier_right<'a>(
+        x_left: &'a BitArray32,
+        y_left: &'a BitArray32,
+        prss_left: &'a BitArray32,
+    ) -> impl Iterator<Item = SingleUVPolynomial<Self>>;
 }
 
-// TODO(dm) - implement Basefield for Fp61BitPrime in follow up PR
+impl DZKPBaseField for Fp61BitPrime {
+    const INVERSE_OF_TWO: Self = Fp61BitPrime::const_truncate(1_152_921_504_606_846_976u64);
+    const MINUS_ONE_HALF: Self = Fp61BitPrime::const_neg(Self::INVERSE_OF_TWO);
+    const MINUS_TWO: Self = Fp61BitPrime::const_neg(Fp61BitPrime::const_truncate(2));
+
+    // Convert allows to convert individual bits from multiplication gates into dzkp compatible field elements
+    // We use the conversion logic from https://eprint.iacr.org/2023/909.pdf
+    //
+    // This function does not use any optimization.
+    //
+    // Prover and left can compute:
+    // g1=-2ac(1-2e),
+    // g2=c(1-2e),
+    // g3=a(1-2e),
+    // g4=-1/2(1-2e),
+    //
+    // Prover and right can compute:
+    // h1=bd(1-2f),
+    // h2=d(1-2f),
+    // h3=b(1-2f),
+    // h4=1-2f,
+    //
+    // where
+    // (a,b,c,d,f) = (xleft, yright, yleft, xright, prssright)
+    // and
+    // e = ab⊕cd⊕ f
+    fn convert_prover<'a>(
+        x_left: &'a BitArray32,
+        x_right: &'a BitArray32,
+        y_left: &'a BitArray32,
+        y_right: &'a BitArray32,
+        prss_right: &'a BitArray32,
+    ) -> impl Iterator<Item = UVPolynomials<Fp61BitPrime>> {
+        x_left
+            .chunks(RECURSION_CHUNK_SIZE)
```

Comment:
I can't seem to import all of these dependencies into the Rust playground to play around with it. When you call:
`x_left.chunks(RECURSION_CHUNK_SIZE)` what is the type of element you receive in the iterator?

Is it a slice of RECURSION_CHUNK_SIZE bit? or is it a slice of RECURSION_CHUNK_SIZE *bytes*?

Reply:
chunks is a function in the bitvec:
https://docs.rs/bitvec/1.0.1/bitvec/slice/struct.BitSlice.html#method.chunks

it iterates over Recursion_chunk_size many bits

Reply:
Since we do not need to support any arbitrary amount of chunked iteration, and we can easily restrict ourselves to iterating by multiples of 8, or 16, or 32, I don't think we need to use a bitvec. We can just use a vec of u8 or u16 or u32. I don't think we are actually utilizing any of the power of bitvec here. Am I missing something? I just feel like it's making the code messier without any additional upside.

Reply:
Yes, I agree that we don't need to support any chunk size. However, for lambda = 8, chunk size would be 2, for lambda = 32, chunk size is 8. We still need to iterate over all the bits, a vec of u8, u16, u32 do not natively support that and I would need to implement my own iterator over bits.